hicommonwealth/commonwealth
# Cosmos node liveness

## Link to Issue
Closes: #5157 

## Description of Changes
- Creates a fallback strategy for failing Cosmos nodes
- Adds a `health` column to ChainNodes
- Renames cosmosLCD to cosmosAPI/v1
   
## "How We Fixed It"

- If a chain node fails, we try the community proxy endpoint provided by cosmos.directory
- For a given period (default 5 minutes), we continue to use the proxy (using health flag and last_updated)

## Test Plan
- Add playwright tests to demonstrate node-failure case (local only)
- Devnet tests will run in CI
- CA (click around) tested on local:
  - http://localhost:8080/ux/discussions/proposals -> Note loaded proposals
  - Set your local db to a failing node for /ux:
```
update "ChainNodes" set alt_wallet_url = 'willfail' where cosmos_chain_id = 'ux';
update "ChainNodes" set url = 'willalsofail' where cosmos_chain_id = 'ux';
```
  - http://localhost:8080/ux/discussions/proposals again -> Expect proposals to still load

## Deployment Plan

1.  Optional env vars added for configuration. Not needed for deployment, although we may want to update the proxy referer when a whitelist header is agreed upon.
`COSMOS_PROXY_REFERER`
`FALLBACK_NODE_DURATION_S` in seconds (defaults to 300 if not provided)

## Other Considerations

- if #6097 is merged first, we will want to make some adjustments to this PR.
